### ISSUE TITLE
Mazinger Z (world) netplay issue (disconnection)

### ISSUE THREAD

**Issue by furiadeoso** [libretro/fbalpha#74]
Symptom:

In 2 players netplay, when both players are playing (If none touch controls, 5 minutes may pass without anything happening), after a seemingly random time but before 30 seconds, a player (can be the host) crashes and retroarch closes.

Nature:
I can not specify the nature of the issue, but it could be related to something slightly wrong in savestates, something related to the inputs and the savestate.

How to reproduce:
The issue can be reproduced in local opening two windows and connecting. Start a game and move, fire, etc. 

Disabling check frames (=0) don't solve the problem. 

Tell me if I can do something more to test.

Regards.

**Comment by barbudreadmon** [comment]
That's a weird one, it seems savestates aren't at fault (couldn't reproduce this with rewind), at the very least not in an obvious way, with local netplay i get a lot of `ZetTotalCycles called when no CPU open` messages, before a crash with the following trace : 
```
==9094==ERROR: AddressSanitizer: global-buffer-overflow on address 0x7f5c57cce2e0 at pc 0x7f5c542026ea bp 0x7ffd945fb1d0 sp 0x7ffd945fb1c0
WRITE of size 4 at 0x7f5c57cce2e0 thread T0
    #0 0x7f5c542026e9 in mazingerZOut src/burn/drv/cave/d_mazinger.cpp:263
    #1 0x7f5c5541de60 in ZetWriteIO(unsigned int, unsigned char) src/cpu/z80_intf.cpp:65
    #2 0x7f5c55625b1c in op_d3 src/cpu/z80/z80.cpp:3281
    #3 0x7f5c5562b00b in Z80Execute(int) src/cpu/z80/z80.cpp:3598
    #4 0x7f5c5541ebd8 in ZetRun(int) src/cpu/z80_intf.cpp:397
    #5 0x7f5c542020b2 in mazingerWriteWord src/burn/drv/cave/d_mazinger.cpp:178
    #6 0x7f5c55429f4b in WriteWord src/cpu/m68000_intf.cpp:357
    #7 0x7f5c55429f4b in M68KWriteWord src/cpu/m68000_intf.cpp:688
    #8 0x7f5c53d0a8a2 in m68ki_write_16_fc src/cpu/m68k/m68kcpu.h:1160
    #9 0x7f5c53d0a8a2 in m68k_op_move_16_al_d src/cpu/m68k/m68kops.c:20236
    #10 0x7f5c53c41c35 in m68k_execute src/cpu/m68k/m68kcpu.c:830
    #11 0x7f5c5542adb5 in SekRun(int) src/cpu/m68000_intf.cpp:1406
    #12 0x7f5c54203123 in DrvFrame src/burn/drv/cave/d_mazinger.cpp:478
    #13 0x7f5c53e89965 in BurnDrvFrame src/burn/burn.cpp:740
    #14 0x7f5c5565b074 in ForceFrameStep src/burner/libretro/libretro.cpp:488
    #15 0x7f5c5565d434 in retro_run src/burner/libretro/libretro.cpp:892
    #16 0x5592c07490ff  (/usr/bin/retroarch+0x470ff)
    #17 0x5592c074ebc7  (/usr/bin/retroarch+0x4cbc7)
    #18 0x5592c0747e3f  (/usr/bin/retroarch+0x45e3f)
    #19 0x7f5c6ec68ae6 in __libc_start_main (/lib64/libc.so.6+0x21ae6)
    #20 0x5592c0744a09  (/usr/bin/retroarch+0x42a09)

0x7f5c57cce2e0 is located 32 bytes to the left of global variable 'SoundLatch' defined in 'src/burn/drv/cave/d_mazinger.cpp:34:14' (0x7f5c57cce300) of size 4
0x7f5c57cce2e0 is located 0 bytes to the right of global variable 'SoundLatchReply' defined in 'src/burn/drv/cave/d_mazinger.cpp:35:14' (0x7f5c57cce220) of size 192
SUMMARY: AddressSanitizer: global-buffer-overflow src/burn/drv/cave/d_mazinger.cpp:263 in mazingerZOut
Shadow bytes around the buggy address:
  0x0fec0af91c00: 00 00 f9 f9 f9 f9 f9 f9 00 00 f9 f9 f9 f9 f9 f9
  0x0fec0af91c10: 00 00 00 00 01 f9 f9 f9 f9 f9 f9 f9 01 f9 f9 f9
  0x0fec0af91c20: f9 f9 f9 f9 01 f9 f9 f9 f9 f9 f9 f9 04 f9 f9 f9
  0x0fec0af91c30: f9 f9 f9 f9 04 f9 f9 f9 f9 f9 f9 f9 04 f9 f9 f9
  0x0fec0af91c40: f9 f9 f9 f9 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0fec0af91c50: 00 00 00 00 00 00 00 00 00 00 00 00[f9]f9 f9 f9
  0x0fec0af91c60: 04 f9 f9 f9 f9 f9 f9 f9 00 f9 f9 f9 f9 f9 f9 f9
  0x0fec0af91c70: 00 f9 f9 f9 f9 f9 f9 f9 01 f9 f9 f9 f9 f9 f9 f9
  0x0fec0af91c80: 01 f9 f9 f9 f9 f9 f9 f9 01 f9 f9 f9 f9 f9 f9 f9
  0x0fec0af91c90: 01 f9 f9 f9 f9 f9 f9 f9 01 f9 f9 f9 f9 f9 f9 f9
  0x0fec0af91ca0: 01 f9 f9 f9 f9 f9 f9 f9 00 f9 f9 f9 f9 f9 f9 f9
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==9094==ABORTING
```
Maybe @dinkc64 can figure out what this is about ? Some remainder about this feature : netplay is handled through savestates streaming in retroarch (basically the ggpo way).

**Comment by barbudreadmon** [comment]
Also linking https://docs.libretro.com/tech/netplay/ , but not sure it'll be helpful

**Comment by dinkc64** [comment]
Hi, please give the wip code I checked in a try, also, see if *san comes up with anything.
I can't find where the last "ZetTotalCycles called when no CPU open" is coming from, this is very strange, because it is printing this outside of the frame and outside of the scan functions..  Maybe there's another oob thing going on?

**Comment by dinkc64** [comment]
Its not related, but it bothers me. :)
sure, close it!

nOpenedCPU doesn't need to be saved, because all cpus are closed when making a state - states happen outside of the frame, and cpu's should only be opened in the frame.
Silly story:
Many years ago when I first started working with fbalpha, I would start scanning every variable I could to make things work, but this turned out to cause a lot more problems than it fixed.  It's really important to know how/why a variable is used before scanning it.
The fbalpha states were a nightmare when I joined.  They didn't work in 9 out of 10 games.. eek :)

best regards,
- dink
